main/src/ca/uwaterloo/flix/util/collection/Chain.scala

```diff
@@ -117,23 +113,19 @@ sealed trait Chain[+A] {
   /**
     * Returns `this` as a [[List]].
     */
-  final def toList: List[A] = this match {
-    // N.B.: We have to use reflection to avoid
-    // infinite recursion when pattern matching
-    // since it calls the equals method which
-    // depends on toList.
-    case _: Chain.Empty.type => List.empty
-    case c: Chain.Link[A] => c.l.toList ++ c.r.toList
-    case c: Chain.Proxy[A] => c.xs.toList
+  final def toList: List[A] = {
+    val buf = new ListBuffer[A]
+    this.foreach(buf.addOne)
+    buf.toList
   }
 
   /**
     * Displays all elements of this collection in a string using a separator string.
     */
-  final def mkString(sep: String): String = this match {
-    case Chain.Empty => ""
-    case Chain.Link(l, r) => l.mkString(sep) ++ sep ++ r.mkString(sep)
-    case Chain.Proxy(xs) => xs.mkString(sep)
+  final def mkString(sep: String): String = {
+    val sb = new StringBuilder()
+    this.toList.addString(sb, sep)
```

Comment:
This looks strange to me. I would have expected a for loop. What is addString on a list? Is that some kind of implicit coercion? I would avoid that.

Reply:
I looked at the docs and just converts each element in the list to a stringbuilder using `sep` to separate the strings. Nothing fancy honestly.
https://www.scala-lang.org/api/current/scala/collection/immutable/List.html#addString-ff4

Reply:
What a weird function name

Reply:
lets not use this function, its very confusing